app/master/build_request_handler.py

```diff
@@ -73,6 +75,22 @@ def _build_preparation_loop(self):
         """
         while True:
             build = self._request_queue.get()
+            project_id = build.project_type.project_id()
+
+            if project_id not in self._project_preparation_locks:
+                self._project_preparation_locks[project_id] = Lock()
+
+            project_lock = self._project_preparation_locks[project_id]
```

Comment:
Because this method (_build_preparation_loop()) cannot be called concurrently, it's safe to perform operations such as this without worrying about concurrency.